### PR TITLE
Fix: sparse_csr_tensor segfaults when crow_indices or col_indices are non-tensors

### PR DESCRIPTION
**Pull request by aocsa**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56561 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr
* #56560 CUDA support in the CSR layout: constructors
* #56744 Fix: Avoid unnecessary creation of tensors in SparseCsrTensorImpl::resize_and_clear_
* **#56723 Fix: sparse_csr_tensor segfaults when crow_indices or col_indices are non-tensors**
* #56722 Fix: Incorrect example output in sparse_csr_tensor doc-string

Fix  gh-56687

Differential Revision: [D27999919](https://our.internmc.facebook.com/intern/diff/D27999919)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -605,13 +605,13 @@ Tensor sparse_csr_tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scal
...
   THPObjectPtr crow_indices_dtype_attr(PyObject_GetAttrString(r.pyobject(CROW_INDICES_ARG), "dtype"));
   THPObjectPtr col_indices_dtype_attr(PyObject_GetAttrString(r.pyobject(COL_INDICES_ARG), "dtype"));
-  at::ScalarType crow_indices_scalar_type = reinterpret_cast<THPDtype*>(
-    crow_indices_dtype_attr.get())->scalar_type;
-  at::ScalarType col_indices_scalar_type = reinterpret_cast<THPDtype*>(
-    col_indices_dtype_attr.get())->scalar_type;
+  at::ScalarType crow_indices_scalar_type = crow_indices_dtype_attr ? reinterpret_cast<THPDtype*>(
+    crow_indices_dtype_attr.get())->scalar_type : kInt;
+  at::ScalarType col_indices_scalar_type = col_indices_dtype_attr ? reinterpret_cast<THPDtype*>(
+    col_indices_dtype_attr.get())->scalar_type : kInt;
```
Comment (aocsa) [inline torch/csrc/utils/tensor_new.cpp:611]: I updated these lines of codes in order to completely fix the segfault issue reported here: https://github.com/pytorch/pytorch/issues/56687

Let me know your thoughts @pearu

Comment (pearu) [reply torch/csrc/utils/tensor_new.cpp:611]: @aocsa Could you implement also tests for it?

**Review by pearu — CHANGES_REQUESTED** [review]
Looks good @aocsa !

It would be nice to have some tests for constructing CSR tensor from non-tensor indices as well.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (aocsa) [reply torch/csrc/utils/tensor_new.cpp:611]: I have the test ready.
```python
    @dtypes(*torch.testing.get_all_dtypes(include_bool=False, include_half=False,
                                          include_bfloat16=False, include_complex=False))
    def test_factory_device_type_inference(self, device, dtype):
        sparse = torch.sparse_csr_tensor([0, 2, 4],
                                         [0, 1, 0, 1],
                                         [1, 2, 3, 4],
                                         size=(2, 10),
                                         dtype=dtype,
                                         device=device)

        self.assertEqual((2, 10), sparse.shape)
        self.assertEqual(torch.tensor([0, 2, 4], dtype=torch.int32, device=device), sparse.crow_indices())
        self.assertEqual(torch.tensor([0, 1, 0, 1], dtype=torch.int32, device=device), sparse.col_indices())
        self.assertEqual(torch.tensor([1, 2, 3, 4], dtype=dtype, device=device), sparse.values())

        cpu_cuda = ('cpu', 'cuda')
        cpu_cuda_none = cpu_cuda + (None,)
        for indices_device, values_device, device in itertools.product(cpu_cuda,
                                                                       cpu_cuda,
                                                                       cpu_cuda_none):
            for index_dtype in [torch.int32, torch.int64]:
                crow_indices = torch.tensor([0, 2, 4], dtype=index_dtype, device=indices_device)
                col_indices = torch.tensor([0, 1, 0, 1], dtype=index_dtype, device=indices_device)
                values = torch.tensor([1, 2, 3, 4], dtype=dtype, device=values_device)
                if device is None and indices_device != values_device:
                    with self.assertRaises(RuntimeError):
                        torch.sparse_csr_tensor(crow_indices,
                                                col_indices,
                                                values,
                                                size=(2, 10),
                                                device=device)
                else:
                    t = torch.sparse_csr_tensor(crow_indices,
                                                col_indices,
                                                values,
                                                size=(2, 10),
                                                device=device)
                    should_be_cuda = (device == 'cuda' or (device is None and values_device == 'cuda'))
                    self.assertEqual(should_be_cuda, t.is_cuda)
```
However, for simplicity and sanity I will add the test once PR #56392 is merged.

**Comment by mruberry** [comment]
Hey @aocsa, would you ping me on this once @pearu's satisfied?

**Review by pearu — APPROVED** [review]
LGTM, it's fine with me that tests come after #56392.
^ @mruberry

**Pull request closed without merging**